### PR TITLE
Fix: Exit Codes

### PR DESCRIPTION
**Pull request by Meldiron**
- Can run both in parallel, and in serial.

Manual QA:

All working:
CleanShot 2022-03-24 at 14 21 43

All failing:
CleanShot 2022-03-24 at 14 21 47

40% failing:
CleanShot 2022-03-24 at 14 21 51

### CODE REVIEW

File: src/CLI/Console.php [hunk 1]
```diff
@@ -133,20 +133,24 @@ static public function exit(int $status = 0): void
      */
     static public function execute(string $cmd, string $stdin, string &$stdout, string &$stderr, int $timeout = -1): int
     {
+        $cmd = '( ' . $cmd . ' ) 3>/dev/null ; echo $? >&3';
```
Comment (PineappleIOnic) [inline src/CLI/Console.php:136]: Is there any chance a command with ' inside of it can cause this function to fail? If so could we add escape characters?

**Review by PineappleIOnic — CHANGES_REQUESTED** [review]

File: src/CLI/Console.php (hunk 1 above)
Comment (Meldiron) [reply src/CLI/Console.php:136]: `echo 'Hello World'` worked fine. I think `'`symbols are only part of PHP syntax in this case.

Comment (PineappleIOnic) [reply src/CLI/Console.php:136]: Alright, lgtm

**Review by PineappleIOnic — APPROVED** [review]

File: src/CLI/Console.php [hunk 2]
```diff
@@ -133,20 +133,24 @@ static public function exit(int $status = 0): void
...
         $process = \proc_open(
             $cmd,
-            [['pipe','r'],['pipe','w'],['pipe','w']],
+            [['pipe','r'],['pipe','w'],['pipe','w'],['pipe', 'w']],
             $pipes
         );
         $start = \time();
         $stdout = '';
         $stderr = '';
+        $statusPipe = '';
```
Comment (christyjacob4) [inline src/CLI/Console.php:147]:
```suggestion
        $status = '';
```

lets just call it status to be consistent with the others....

File: src/CLI/Console.php [hunk 3]
```diff
@@ -155,6 +159,7 @@ static public function execute(string $cmd, string $stdin, string &$stdout, stri
         while (\is_resource($process)) {
             $stdout .= \stream_get_contents($pipes[1]);
             $stderr .= \stream_get_contents($pipes[2]);
+            $statusPipe .= \stream_get_contents($pipes[3]);
```
Comment (christyjacob4) [inline src/CLI/Console.php:162]:
```suggestion
            $status .= \stream_get_contents($pipes[3]);
```

File: src/CLI/Console.php [hunk 4]
```diff
@@ -168,7 +173,9 @@ static public function execute(string $cmd, string $stdin, string &$stdout, stri
                 \fclose($pipes[2]);
                 \proc_close($process);
 
-                return (int)$status['exitcode'];
+                $exitCode = (int) str_replace("\n","",$statusPipe);
```
Comment (christyjacob4) [inline src/CLI/Console.php:176]:
```suggestion
                $exitCode = (int) str_replace("\n","",$status);
```

**Review by christyjacob4 — CHANGES_REQUESTED** [review]
Some minor comments.

Can you also explain / share a link to the exact logic here ?

**Comment by Meldiron** [comment]
> Some minor comments.
> 
> Can you also explain / share a link to the exact logic here ?

`'( ' . $cmd . ' ) 3>/dev/null ; echo $? >&3'`

`($cmd)` executes a command
`3` means we are creating a new pipe (stdin, stderr, stdout, OUR NEW ONE)
`>` forwards output of the command
`/dev/null` means we don't display output
`;` indicates a new command
`echo $?` prints the exit code of the last command
`>` again, forwards output of the command
`&3` makes sure to print output to 3rd pipe

Finally, we get output of 3rd pipe in PHP `\stream_set_blocking($pipes[3], false);`.

That's how I remember understanding it. The website was a really old forum, can't find it anymore.

**Review by christyjacob4 — APPROVED** [review]

**Pull request merged**